ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1721,6 +1759,484 @@ void shutdown() throws InterruptedException {
+          GetResponse response = GetResponse.readFrom(dis, clusterMap);
+          ReplicaThread.this.handleGetResponse(response, remoteReplicaInfos, exchangeMetadataResponseList,
+              remoteDataNode, isNonProgressStandbyReplicaGroup);
+          logger.trace(
+              "Remote node: {} Thread name: {} RemoteReplicaGroup {} Handled GetResponse for correlation id {}",
+              dataNodeId, threadName, id, responseInfo.getRequestInfo().getRequest().getCorrelationId());
+          long fixMissingStoreKeysTime = time.milliseconds() - fixMissingStoreKeysStartTimeInMs;
+          replicationMetrics.updateFixMissingStoreKeysTime(fixMissingStoreKeysTime, replicatingFromRemoteColo,
+              replicatingOverSsl, datacenterName);
+        } catch (Exception e) {
+          setException(e, "GetResponse deserialization received unexpected error");
+        }
+      } else {
+        remoteReplicaInfos.forEach(r -> responseHandler.onEvent(r.getReplicaId(), networkClientErrorCode));
+        setException(new IOException("NetworkClientErrorCode: " + networkClientErrorCode), "Failed to send GetRequest");
+      }
+      state = ReplicaGroupReplicationState.DONE;
+    }
+
+    /**
+     * Set exception for the group and change the state to DONE and log the message out.
+     * @param e The exception to set
+     * @param message The message to log out
+     */
+    private void setException(Exception e, String message) {
+      logger.error("Remote node: {} Thread name: {} RemoteReplicaGroup {} {}", remoteDataNode, threadName, id, message,
+          e);
+      exception = e;
+      state = ReplicaGroupReplicationState.DONE;
+    }
+
+    /**
+     * @return The exception.
+     */
+    public Exception getException() {
+      return exception;
+    }
+  }
+
+  /**
+   * Do replication for replicas grouped by {@link DataNodeId} with nonblocking network client.
+   */
+  public void replicateWithNetworkClient() {
+    exchangeMetadataResponsesInEachCycle = new HashMap<>();
+    long oneRoundStartTimeMs = time.milliseconds();
+    logger.trace("Thread name: {} Start RemoteReplicaGroup replication", threadName);
+    List<RemoteReplicaGroup> remoteReplicaGroups = new ArrayList<>();
+    int remoteReplicaGroupId = 0;
+
+    // Before each cycle of replication, we clean up the cache in key converter.
+    storeKeyConverter.dropCache();
+    Map<DataNodeId, List<RemoteReplicaInfo>> dataNodeToRemoteReplicaInfo = getRemoteReplicaInfos();
+    for (Map.Entry<DataNodeId, List<RemoteReplicaInfo>> entry : dataNodeToRemoteReplicaInfo.entrySet()) {
+      DataNodeId remoteNode = entry.getKey();
+      List<RemoteReplicaInfo> replicasToReplicatePerNode = entry.getValue();
+      List<RemoteReplicaInfo> activeReplicasPerNode = new ArrayList<>();
+      List<RemoteReplicaInfo> standbyReplicasWithNoProgress = new ArrayList<>();
+      filterRemoteReplicasToReplicate(replicasToReplicatePerNode, activeReplicasPerNode, standbyReplicasWithNoProgress);
```

Comment:
Probably it's not introduced by this PR.
Looks like for each replication cycle, we have to check replica status and filter active, standbyReplicas and setup groups. It introduced extra overhead.
Ideally only when there is some cluster map change, we may need do that.
Not necessary for this PR. Just something if we can improve.

Reply:
It's not that. We have different logic for standby replicas replicating from remote datacenter. If the local replica is a standby replica, then it doesn't replicate blob content from any replicas in remote datacenter.

For example, in DC1, we have replica A, B, C and in DC2, we have replica X, Y, Z. Replica A and X are the leaders for each DC. If this host has replica B, and it will try to replicate from A, C and X, Y, Z. However, copying blob content cross datacenter is expensive, so we don't allow B to copy blob content, we only allow B to replicate metadata cross datacenter. So B would replicate metadata and data from A and C, but only metadata from X, Y, Z.

This filter logic happens before we send out the GetRequest to replicate data, so we will filter out the RemoteReplica X, Y, Z.

Reply:
My point is that theoretically we don't need do this for every cycle. 
For this host and this partition, we always replicate data from A and C and always filter out X,Y,Z. It never change unless A or C is down and B become the leader.

Reply:
But I guess the code path is not the performance bottleneck yet, it's ok.